### PR TITLE
Fixes issue reverting osgi bom files with SNAPSHOT versions

### PR DESCRIPTION
**Pull request by grkvlt**
Fix `install()` of SNAPSHOT versioned bundles so that checksums are never compared, and bundles are always updated. This is because a SNAPSHOT bundle uses the checksum of the first installation to determine the URL and saves this as metadata, which is never changed.

Signed-off-by: Andrew Donald Kennedy <andrew.kennedy@cloudsoftcorp.com>

### CODE REVIEW

**Review by duncangrant — DISMISSED** [review]
LGTM thanks @grkvlt

**Comment by ahgittin** [comment]
hold fire on this -- failing test?

File: core/src/main/java/org/apache/brooklyn/util/core/osgi/BundleMaker.java [hunk 1]
```diff
@@ -269,7 +269,8 @@ private void writeZipEntriesFromFile(ZipOutputStream zout, File existingZip, Pre
     @Deprecated
     public Bundle installBundle(File f, boolean start) {
         try {
-            Bundle b = Osgis.install( framework, "file://"+f.getAbsolutePath() );
+
+            Bundle b = Osgis.install( framework, f.toURI().toString() );
```
Comment (ahgittin) [inline core/src/main/java/org/apache/brooklyn/util/core/osgi/BundleMaker.java:273]: suspicious

File: core/src/test/java/org/apache/brooklyn/core/config/external/PropertiesFileExternalConfigSupplierTest.java [hunk 2]
```diff
@@ -68,11 +68,6 @@ public void testFromPropertiesJavaFileUrl() throws Exception {
         doTestFromProperties(f -> f.toURI().toString());
     }
 
-    @Test
-    public void testFromPropertiesTwoSlashAndPathUrl() throws Exception {
```
Comment (ahgittin) [inline core/src/test/java/org/apache/brooklyn/core/config/external/PropertiesFileExternalConfigSupplierTest.java:72]: this is a useful test

File: launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java [hunk 3]
```diff
@@ -254,8 +254,8 @@ public void testRebindWithSystemBundleInCatalog() throws Exception {
             // Launch brooklyn, where persisted state now includes the initial catalog's bundle
             startT2(newLauncherForTests(initialBomFile.getAbsolutePath()));
             
-            // Should not have replaced the original "system bundle"
-            assertOnlyBundle(reusedFramework, bundleName, pseudoSystemBundle);
+            // FIXME Should we have replaced the original "system bundle" ???
+            assertBundleExists(reusedFramework, bundleName, pseudoSystemBundle);
```
Comment (ahgittin) [inline launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java:258]: feels wrong

**Comment by ahgittin** [comment]
The problem here is that if we have a snapshot bundle, say `foo:1-SNAPSHOT`, and we add it when it's in state A, then add when in state B, and then try to add state A again.  The last add is ignored because it matches a previous state, when actually it should match the most recent state.

I think this needs specific logic to compare with the checksum of the most recent version of a bundle:version, bail out if those match.  And then extra logic if it matches a previous known state of a bundle:version then it restores that (this. might be a new result code).

This also needs a test.

File: core/src/main/java/org/apache/brooklyn/util/core/osgi/BundleMaker.java (hunk 1 above)
Comment (grkvlt) [reply core/src/main/java/org/apache/brooklyn/util/core/osgi/BundleMaker.java]: removed

File: core/src/test/java/org/apache/brooklyn/core/config/external/PropertiesFileExternalConfigSupplierTest.java (hunk 2 above)
Comment (grkvlt) [reply core/src/test/java/org/apache/brooklyn/core/config/external/PropertiesFileExternalConfigSupplierTest.java]: reinserted

File: launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java (hunk 3 above)
Comment (grkvlt) [reply launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java]: reverted test

**Comment by grkvlt** [comment]
FYI, there is another way of doing this, which involves updating the metadata with the new checksum and URL each time, but this caused a test failure and also had the result of keeping old versions around but leaving them unmanaged and confusing brooklyn and osgi, so i gave up on that eventually ;(

**Comment by ahgittin** [comment]
good job identifying the problem.  re solution we _do_ have to update the bundle metadata, otherwise we keep installing snapshots even when they are unchanged which is disruptive.  re-installation should be idempotent.  added a test for that and a fix for it in https://github.com/apache/brooklyn-server/pull/1176 .  merged this and that.  but @grkvlt if you get the chance to eyeball that PR post-merge that would be great.

**Pull request closed without merging**